src/components/AppNavigator/AppNavigator.tsx

```diff
@@ -1,44 +1,117 @@
-import { useEffect } from 'react';
+import { useContext, useEffect } from 'react';
 import { Link as RouterLink, useLocation, useNavigate } from 'react-router-dom';
 import Link from '@mui/material/Link';
 import Step from '@mui/material/Step';
 import Stepper from '@mui/material/Stepper';
-import StepContent from '@mui/material/StepContent';
 import StepLabel from '@mui/material/StepLabel';
-import MenuItemType from 'types/menuItem';
+import Typography from '@mui/material/Typography';
+import FormScreens from 'constants/formScreens';
+import routes from 'constants/routes';
+import WizardFormContext from 'contexts/WizardFormContext';
+import Entries from 'types/entries';
+import { FormStepsList, FormValuesType } from 'types/formTypes';
 
 interface Props {
-  list: MenuItemType[];
-  firstUncompletedStep: number;
+  className?: string;
 }
 
-export default function AppNavigator({ list, firstUncompletedStep }: Props) {
+export default function AppNavigator({ className }: Props) {
+  const { formValues, firstUncompletedStep } = useContext(WizardFormContext);
   const { pathname } = useLocation();
   const navigate = useNavigate();
+  // https://www.charpeni.com/blog/properly-type-object-keys-and-object-entries
+  const formValuesEntries = Object.entries(
+    formValues,
+  ) as Entries<FormValuesType>;
 
   useEffect(() => {
-    if (pathname !== list[firstUncompletedStep].url) {
-      navigate(list[firstUncompletedStep].url);
+    const formScreen = (Object.keys(routes) as (keyof typeof routes)[]).find(
+      (key) => routes[key] === pathname,
+    );
+    const findStep = () => {
+      // eslint-disable-next-line no-restricted-syntax
+      for (const [screenName, step] of formValuesEntries) {
+        if (screenName === formScreen) {
+          return step;
+        }
+        if ('subStep' in step && !!step.subStep) {
+          return step.subStep;
+        }
+      }
+      throw new Error('Unknown step/substep');
+    };
+
+    const currentStep = findStep();
+    if (!currentStep.isCompleted && formScreen !== firstUncompletedStep) {
+      navigate(routes[firstUncompletedStep]);
     }
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [list, pathname]);
+  }, [firstUncompletedStep, pathname, navigate]);
```

Comment:
Add, please, `formValueEntries` and test implementation one more time. I think we can eliminate disabling of ESLint